src/Adapter/Product/CommandHandler/GenerateProductCombinationsHandler.php

```diff
@@ -0,0 +1,185 @@
+ * Handles @see GenerateProductCombinationsCommand using legacy object model
+ */
+final class GenerateProductCombinationsHandler extends AbstractProductHandler implements GenerateProductCombinationsHandlerInterface
+{
+    /**
+     * @var CombinationGeneratorInterface
+     */
+    private $combinationGenerator;
+
+    /**
+     * @var Db
+     */
+    private $dbInstance;
+
+    /**
+     * @param CombinationGeneratorInterface $combinationGenerator
+     */
+    public function __construct(
+        CombinationGeneratorInterface $combinationGenerator
+    ) {
+        $this->combinationGenerator = $combinationGenerator;
+        $this->dbInstance = Db::getInstance();
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @todo: multistore
+     */
+    public function handle(GenerateProductCombinationsCommand $command): array
+    {
+        $product = $this->getProduct($command->getProductId());
+        $generatedCombinations = $this->combinationGenerator->generate($command->getGroupedAttributeIds());
+
+        SpecificPriceRule::disableAnyApplication();
+
+        $combinationIds = $this->addCombinations($product, $generatedCombinations);
+
+        Product::updateDefaultAttribute($product->id);
+        SpecificPriceRule::enableAnyApplication();
+        SpecificPriceRule::applyAllRules([$product->id]);
+
+        return $combinationIds;
+    }
+
+    /**
+     * @param Product $product
+     * @param Iterator $generatedCombinations
+     *
+     * @return CombinationId[]
+     *
+     * @throws CombinationException
+     */
+    private function addCombinations(Product $product, Iterator $generatedCombinations): array
+    {
+        $addedCombinationIds = [];
+        foreach ($generatedCombinations as $generatedCombination) {
+            $combinationExists = $product->productAttributeExists($generatedCombination, false, null, true);
+
+            if ($combinationExists) {
+                continue;
+            }
+
+            try {
+                $addedCombinationIds[] = $this->addCombination($product, $generatedCombination);
+            } catch (PrestaShopException $e) {
+                throw new CombinationException(
+                    sprintf(
+                        'Error occurred when trying to add combination for product #%d. Combination: %s',
+                        $product->id,
+                        var_export($generatedCombination, true)
+                    ),
+                    $e
+                );
+            }
+        }
+
+        return $addedCombinationIds;
+    }
+
+    /**
+     * @param Product $product
+     * @param int[] $generatedCombination
+     *
+     * @return CombinationId
+     */
+    private function addCombination(Product $product, array $generatedCombination): CombinationId
+    {
+        $newCombination = new Combination();
+        $newCombination->id_product = $product->id;
+        $newCombination->default_on = 0;
+
+        $this->dbInstance->execute('START TRANSACTION');
```

Comment:
Maybe we should add methods into the DB object instead, it would be cleaner

Also instead of directly executing the transaction it would be better to use the driver sub methods (for PDO, mysqli, ...) Because they also include some test about nested requests and so on, let's take profit of the expertise of existing code :)

Reply:
:heavy_check_mark: